File: sunnad-ios/sunnad-ios/Domain/Models/Weekday.swift
import Foundation

enum Weekday: Int, CaseIterable, Codable, Hashable, Sendable {
    case monday = 1
    case tuesday = 2
    case wednesday = 3
    case thursday = 4
    case friday = 5
    case saturday = 6
    case sunday = 7

    nonisolated static func fromISOWeekday(_ value: Int) -> Weekday? {
        Weekday(rawValue: value)
    }

    nonisolated static func isoWeekday(for date: Date, calendar: Calendar = .current, timeZone: TimeZone = .current) -> Weekday {
        var calendar = calendar
        calendar.timeZone = timeZone
        let weekday = calendar.component(.weekday, from: date)
        // Calendar weekday: 1=Sun...7=Sat. Convert to ISO: 1=Mon...7=Sun.
        let iso = ((weekday + 5) % 7) + 1
        return Weekday(rawValue: iso) ?? .monday
    }

    nonisolated static func fromMondayFirstIndex(_ index: Int) -> Weekday? {
        Weekday(rawValue: index + 1)
    }

    nonisolated var mondayFirstIndex: Int {
        rawValue - 1
    }
}

File: sunnad-android/app/src/main/java/com/arystan/almasuly/sunnadandroid/core/model/Weekday.kt
package com.arystan.almasuly.sunnadandroid.core.model

import java.time.DayOfWeek
import java.time.LocalDate

enum class Weekday(val isoValue: Int) {
    MONDAY(1),
    TUESDAY(2),
    WEDNESDAY(3),
    THURSDAY(4),
    FRIDAY(5),
    SATURDAY(6),
    SUNDAY(7);

    val mondayFirstIndex: Int
        get() = isoValue - 1

    companion object {
        fun fromIsoWeekday(value: Int): Weekday? = entries.firstOrNull { it.isoValue == value }

        fun isoWeekday(date: LocalDate): Weekday {
            val iso = date.dayOfWeek.value
            return fromIsoWeekday(iso) ?: MONDAY
        }

        fun fromMondayFirstIndex(index: Int): Weekday? = fromIsoWeekday(index + 1)

        fun fromDayOfWeek(dayOfWeek: DayOfWeek): Weekday = fromIsoWeekday(dayOfWeek.value) ?: MONDAY
    }
}
